src/acl/ConnMark.cc

```diff
@@ -19,99 +19,87 @@
-            _mark = static_cast<nfmark_t>(tmp_result);
+        if (tokenizer.int64(number, 0, false))
+            mark = static_cast<nfmark_t>(number);
         else {
-            debugs(28, DBG_CRITICAL, MYNAME << ": bad mark '" << token << "'");
-            self_destruct();
+            throw TexcHere(ToSBuf("acl ", typeString(), ": bad mark value ", token));
         }
 
-        // mark with mask ?
-        SBuf::size_type pos = token.find('/');
-        if (pos != SBuf::npos) {
-            token = SBuf(t + pos + 1);
-            tokenizer = Parser::Tokenizer(token);
-            if (tokenizer.int64(tmp_result, 0, false))
-                _mask = static_cast<nfmark_t>(tmp_result);
-            else {
-                debugs(28, DBG_CRITICAL, MYNAME << ": bad mask '" << token << "'");
-                self_destruct();
+        if (!tokenizer.atEnd()) {
+            if (tokenizer.skip('/')) {
+                if (tokenizer.int64(number, 0, false))
+                    mask = static_cast<nfmark_t>(number);
+                else {
+                    throw TexcHere(ToSBuf("acl ", typeString(), ": bad mask value ", token));
+                }
+                if (!tokenizer.atEnd()) {
+                    throw TexcHere(ToSBuf("acl ", typeString(), ": bad mask value - garbage after number ", token));
+                }
+            } else {
+                throw TexcHere(ToSBuf("acl ", typeString(), ": bad mark value - garbage after number ", token));
             }
         }
-
-        marks.insert(std::pair<nfmark_t, nfmark_t>(_mark, _mask));
-        debugs(28, DBG_DATA, MYNAME << ": mark '" << _mark << "/" << _mask << "'");
+        marks.insert(std::pair<nfmark_t, nfmark_t>(mark, mask));
+        debugs(28, 7, "mark '" << mark << "/" << mask << "'");
     }
 
     if (marks.empty()) {
-        debugs(28, DBG_CRITICAL, MYNAME  << ": expect at least one connmark");
-        self_destruct();
+        throw TexcHere(ToSBuf("acl ", typeString(), " requires at least one mark"));
     }
 }
 
 int
-ACLConnMark::match(ACLChecklist *cl)
+Acl::ConnMark::match(ACLChecklist *cl)
 {
-    ACLFilledChecklist *checklist = Filled(cl);
-    nfmark_t conn_mark = checklist->conn()->clientConnection->nfmark;
-
-    if (marks.empty())
-        return -1;
+    const auto *checklist = Filled(cl);
+    nfmark_t connMark = checklist->conn()->clientConnection->nfmark;
 
-    for (auto m : marks) {
-        if ((conn_mark & m.second) == m.first) {
+    for (const auto &m : marks) {
+        if ((connMark & m.second) == m.first) {
+            debugs(28, 7, "CONNMARK '" << connMark << "' matches with '" << m.first <<  "/" << m.second << "'");
             return 1;
         }
     }
     return 0;
 }
 
 SBufList
-ACLConnMark::dump() const
+Acl::ConnMark::dump() const
 {
     SBufList sl;
-    for (auto m : marks) {
+    for (const auto &m : marks) {
         SBuf s;
-        s.Printf("0x%08x/0x%08x", m.first, m.second);
+        s.Printf("0x%08x", m.first);
```

Comment:
I suspect you will find writing good debugging statements, exceptions, and this dump() code easier if you add an operator to print `std::pair<nfmark_t, nfmark_t>` into std::ostream!